### PR TITLE
8302987: Add uniform and spatially equidistributed bounded double streams to RandomGenerator

### PR DESCRIPTION
**Pull request by rgiulietti**
The `default` method `nextDouble(double origin, double bound)` in `java.util.random.RandomGenerator` aims at generating a uniformly and spatially equidistributed random `double` in the left-closed and right-open range [`origin`, `bound`). It does so by applying the affine transform `origin + (bound - origin) * r` to a uniformly and spatially equidistributed random `double` `r` in the range [0, 1).

Since floating-point arithmetic suffers from small but noticeable rounding errors, this ends up slightly deforming the distribution of `r` when applying the affine transform.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8303584](https://bugs.openjdk.org/browse/JDK-8303584) to be approved

### Issues
 * [JDK-8302987](https://bugs.openjdk.org/browse/JDK-8302987): Add uniform and spatially equidistributed bounded double streams to RandomGenerator (**Enhancement** - P4)
 * [JDK-8303584](https://bugs.openjdk.org/browse/JDK-8303584): Add uniform and spatially equidistributed bounded double streams to RandomGenerator (**CSR**)

### Reviewers
 * [Joe Darcy](https://openjdk.org/census#darcy) (@jddarcy - **Reviewer**) ⚠️ Review applies to [80e49d41](https://git.openjdk.org/jdk/pull/12719/files/80e49d41c58b17d5da9b5ef9aaed5227949902ba)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12719/head:pull/12719` \
`$ git checkout pull/12719`

Update a local copy of the PR: \
`$ git checkout pull/12719` \
`$ git pull https://git.openjdk.org/jdk.git pull/12719/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12719`

View PR using the GUI difftool: \
`$ git pr show -t 12719`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12719.diff">https://git.openjdk.org/jdk/pull/12719.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12719#issuecomment-1440263522)

### CODE REVIEW

**Comment by rgiulietti** [comment]
The effect of rounding errors in the affine transform is analyzed in:
Goualard, "Drawing random floating-point numbers from an interval", ACM Transactions on Modeling and Computer Simulation, 2022, 32 (3) available [here] (https://hal.science/hal-03282794v4)

The code in this PR is inspired by that paper, although it proposes another implementation preserving uniformity and equidistribution.

**Comment by rgiulietti** [comment]
A similar implementation for `float`s will be added once this PR has elicited some interest and the algorithm has undergone a first review.

A CSR and tests will then be added as well.

**Review by jddarcy — APPROVED** [review]
I read over the cited paper and the code in the PR.

If this new method is sufficiently used, we could considering adding and equidoubles(double a, b) overload for the [a, b) case.

File: src/java.base/share/classes/java/util/random/RandomGenerator.java [hunk 1]
```diff
@@ -252,6 +253,169 @@ default DoubleStream doubles(long streamSize, double randomNumberOrigin,
...
+             * Similar considerations show that kr is the smallest integer such
+             * that kr delta lies to the right of I (it is an exact double).
+             *
+             * All the above means that (kl + nextLong(n)) * delta is an exact
+             * double lying in I and that kl and kr, thus n, are the best
+             * possible choices to ensure the largest number of equidistant
+             * doubles in I. Uniform distribution relies on the guarantee
+             * afforded by nextLong().
+             */
+            delta = nextUp(left) - left;
```
Comment (jddarcy) [inline src/java.base/share/classes/java/util/random/RandomGenerator.java:398]: The delta computations are equivalent to some expression phrased in terms of an ulp (Math.ulp) of a the endpoint or a value adjacent to the endpoint. If the ulp method is not used for the computation, I suggest adding a comment noting the equivalence.

Comment (rgiulietti) [reply src/java.base/share/classes/java/util/random/RandomGenerator.java:398]: Here `left` is negative or `0.0`. What is needed is the distance to the next adjacent `double` in the direction of positive infinity. This is almost always the same as `Math.ulp(left)`, but not always.
For example, when `left` is `-1.0`, `delta` as computed here is half of `Math.ulp(left)`.

Analogously for the case covered by the `else` arm.

**Pull request closed without merging**